core/src/main/scala/org/apache/spark/sql/delta/commands/ShowDeltaPartitionsCommand.scala

```diff
@@ -0,0 +1,124 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package org.apache.spark.sql.delta.commands
+
+import org.apache.hadoop.fs.Path
+import org.apache.spark.sql.delta.{DeltaErrors, DeltaLog, DeltaTableIdentifier}
+import org.apache.spark.sql.{Row, SparkSession}
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.catalyst.analysis.{NoSuchDatabaseException, NoSuchTableException}
+import org.apache.spark.sql.catalyst.catalog.{CatalogTable, CatalogTableType}
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.delta.metering.DeltaLogging
+import org.apache.spark.sql.execution.command.LeafRunnableCommand
+import org.apache.spark.sql.functions.col
+import org.apache.spark.sql.types.StructType
+
+/**
+ * A command for users to list the partition names of a delta table. If partition spec is specified,
+ * partitions that match the spec are returned. Otherwise an empty result set is returned.
+ */
+case class ShowDeltaPartitionsCommand(
+     path: Option[String],
+     tableIdentifier: Option[TableIdentifier]
+     // partitionKeys: Option[Map[String, String]]
+                                     )
+  extends DeltaCommandWithoutViewsSupport(path, tableIdentifier) with DeltaLogging {
+
+  val PARTITION_VALUES = "partitionValues"
+
+  override val output: Seq[Attribute] = getPartitionSchema.toAttributes
+
+  override def run(spark: SparkSession): Seq[Row] = {
+
+    val basePath = getPathAndTableMetadata(spark, "SHOW PARTITIONS")._1
+    val deltaLog = DeltaLog.forTable(spark, basePath)
+
+    recordDeltaOperation(deltaLog, "delta.ddl.showPartitions") {
+      val projection = deltaLog.snapshot.metadata.partitionSchema.map{ field =>
+        col(PARTITION_VALUES + "." + field.name).cast(field.dataType)
+      }
+      if(projection.isEmpty) {
+        throw DeltaErrors.showPartitionsOnNonPartitionedTableException(basePath.toString)
+      }
+
+      deltaLog.snapshot.allFiles
+        .select(projection: _*)
+        .distinct()
+        .collect()
+        .toSeq
+    }
+  }
+
+  override def viewNotSupportedException(view: TableIdentifier): Throwable = {
+    DeltaErrors.viewInShowPartitionsException(view)
+  }
+
+  protected def getPartitionSchema: StructType = {
+    val spark = SparkSession.getActiveSession.getOrElse {
+      throw new IllegalArgumentException("Could not find active SparkSession")
+    }
+    val basePath = getPathAndTableMetadata(spark, "SHOW PARTITIONS")._1
+    val deltaLog = DeltaLog.forTable(spark, basePath)
+    deltaLog.snapshot.metadata.partitionSchema
+  }
+}
+
+abstract class DeltaCommandWithoutViewsSupport(
```

Comment:
So first, I noticed `DeltaCommand` uses `DeltaErrors.viewNotSupportedException(operationName)`, we should be able to use this error instead (and delete the command specific ones in `DeltaErrors`.)

I also noticed a lot of this code overlaps with `DeltaCommand.getDeltaLog`, which for some reason neither `DescribeDeltaDetailsCommand` nor `DescribeDeltaHistoryCommand` extends. In fact it seems to me `DescribeDeltaHistoryCommand` basically copies exactly `getDeltaLog`.

What I'm currently thinking needs to happen is...

1) `DescribeDeltaHistoryCommand` should extend `DeltaCommand` and use `DeltaCommand.getDeltaLog`
2) We should see if we can combine this code here (as used by `DescribeDeltaDetailsCommand`) with `DeltaCommand.getDeltaLog` to minimize overlap. I'm currently thinking something like ...
```
trait DeltaCommand {
   ...
   def getPathAndTableMetadata(...) {
      ...
   }
   def getDeltaLog(...) {
       path, metadata = getPathAndTableMetadata(...)
       // check if delta table and throw error if so
       ....
   }
```
3) Do we need to support `ShowDeltaPartitionsCommand` for non-delta tables (@zsxwing I'm thinking no)? Right now we aren't checking if it's a Delta table before creating the `DeltaLog`. Assuming we only need to support Delta tables, we should then be able to use `DeltaCommand.getDeltaLog` here since both usages of `getPathAndTableMetadata` ignores `tableMetadata` and only uses `basePath` to create a `DeltaLog`.

(1) and (2) should be done in a separate PR. I think in this PR we can just revert the changes to `DescribeDeltaDetailsCommand`, and use `DeltaCommand.getDeltaLog`.

We should also update `DeltaCommand.getDeltaLog` to throw the more helpful error for temporary views, as we have done in `DescribeDeltaDetailCommand` (good reason for reducing code duplication 😄 ).

What do you think @zsxwing?

Reply:
LMK if you'd be willing to tackle 1&2 in a separate PR.